engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,660 @@
+		return errf(
+			fmt.Sprintf(
+				"cannot update stopHeight, stopping commenced for %s",
+				s.stopBoundary),
+		)
+	}
+
+	if !s.stopBoundary.Set() {
+		// if the current stop is no stop, we can always update
+		return nil
+	}
+
+	if s.stopBoundary.fromVersionBeacon == newStopBoundary.fromVersionBeacon {
+		// if the stop was set by the same source, we can always update
+		return nil
+	}
+
+	// if one stop was set by the version beacon and the other one was manual
+	// we can only update if the new stop is strictly earlier
+	if newStopBoundary.StopBeforeHeight < s.stopBoundary.StopBeforeHeight {
+		return nil
+
+	}
+	// this prevents users moving the stopHeight forward when a version newStopBoundary
+	// is earlier, and prevents version beacons from moving the stopHeight forward
+	// when a manual stop is earlier.
+	return errf("cannot update stopHeight, " +
+		"new stop height is later than the current one")
+}
+
+// GetStopParameters returns the upcoming stop parameters or nil if no stop is set.
+func (s *StopControl) GetStopParameters() StopParameters {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.stopBoundary.StopParameters
+}
+
+// ShouldExecuteBlock should be called when new block can be executed.
+// The block should not be executed if its height is above or equal to
+// s.stopBoundary.StopBeforeHeight.
+//
+// It returns a boolean indicating if the block should be executed.
+func (s *StopControl) ShouldExecuteBlock(b *flow.Header) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	// don't process anymore blocks if stopped
+	if s.stopped {
+		return false
+	}
+
+	// Skips blocks at or above requested stopHeight
+	// doing so means we have started the stopping process
+	if b.Height < s.stopBoundary.StopBeforeHeight {
+		return true
+	}
+
+	s.log.Info().
+		Msgf("Skipping execution of %s at height %d"+
+			" because stop has been requested %s",
+			b.ID(),
+			b.Height,
+			s.stopBoundary)
+
+	// stopBoundary is now immutable, because it started affecting execution
+	s.stopBoundary.immutable = true
+	return false
+}
+
+// blockFinalized is called when a block is marked as finalized
+//
+// Once finalization reached stopHeight we can be sure no other fork will be valid at
+// this height, if this block's parent has been executed, we are safe to stop.
+// This will happen during normal execution, where blocks are executed
+// before they are finalized. However, it is possible that EN block computation
+// progress can fall behind. In this case, we want to crash only after the execution
+// reached the stopHeight.
+func (s *StopControl) blockFinalized(
+	ctx irrecoverable.SignalerContext,
+	h *flow.Header,
+) {
+	s.Lock()
+	defer s.Unlock()
+
+	// already stopped, nothing to do
+	if s.stopped {
+		return
+	}
+
+	// We already know the ID of the block that should be executed last nothing to do.
+	// Node is stopping.
+	if s.stopBoundary.stopAfterExecuting != flow.ZeroID {
+		return
+	}
+
+	handleErr := func(err error) {
```

Comment:
a cleaner approach would be return error for such function and raise exception at higher level. Returning with error is the preferred way.